Cargo.toml

```diff
@@ -34,7 +34,9 @@ serde_json = "^1.0.17"
 rmp-serde = "^0.13.7"
 
 [features]
+default = ["serde", "std", "bundled"]
 benchmarks = []
 std = []
-default = ["serde", "std"]
-use-pkg-config = ["libsodium-sys/use-pkg-config"]
+bundled = ["libsodium-sys/cc"]
```

Comment:
I want the first version of this issue to be a simple as possible and with backward compatibility. I don't see any reason why we should change our env flags to features. Could you explain why do you want to do that?

Reply:
> I want the first version of this issue to be a simple as possible and with backward compatibility.

I can certainly keep things fully compatible. I just wanted to start with the untwisted version to the start the discussion on what we would like to have if there were no backwards compatibility concerns. After we clarified that, we can mechanically infer what can we can have with in a compatible way or how much compatibility we are going to loose in exchange for some features.

> I don't see any reason why we should change our env flags to features. 

Environment variables need to be specified every time a build is invoked as @stouset pointed out in #359 as Cargo has no way to specify them persistently. Feature flags however can be specified via a dependent crates Cargo.toml to easily and persistently set a preference for linking with the system library.

Reply:
(For example, the only solution to #359 that is compatible that I can think is accepting both which is messy for people starting to use the crate but can be easily added here.)

Reply:
sodiumoxide is used as a library which means that it can be used multiple times in a binary project which can cause `features` interference. 

> Environment variables need to be specified every time a build is invoked as

People may use `export ENV=value` to set env variable which will live until you `unset` it or end the console session

Reply:
> sodiumoxide is used as a library which means that it can be used multiple times in a binary project which can cause `features` interference.

Yes, features need to be additive. But I tried to preserve this here, i.e. enabling a feature will enable one more way we search for libsodium with system-wide taking precedence as this is what needs to be requested explicitly.

> > Environment variables need to be specified every time a build is invoked as
> 
> People may use `export ENV=value` to set env variable which will live until you `unset` it or end the console session

For a single console session, this helps. But one still needs documentation or script to set up the initial session. This gets more cumbersome if it isn't the top-level crate the requests this. (Both pkg-config/vcpkg provide variables to suppress them on per-library basis which is useful to counteract additive features which seems to mesh well.)

As with the environment variable naming, we can probably support both if we want to (we have a pkg-config feature already), i.e. triggering pkg-config/vcpkg via feature flags and/or via an environment variable.

Reply:
How about fixing `#cfg` issues and then deciding what to do with env vs features in another PR?

Reply:
> How about fixing `#cfg` issues and then deciding what to do with env vs features in another PR?

I can certainly open another pull request try to just fix using up `#[cfg(...)]` usage but that is made difficult because frankly speaking the current `build.rs` is a bit of a mess. Hence the desire to clean up so that the different ways of linking libsodium are spelled our clearly and aligned with where the crates ecosystem as a whole seems to be going.

Reply:
> I can certainly open another pull request

Yes please. This PR is too big and can cause unexpected bugs if not reviewed properly.

Reply:
> Yes please. This PR is too big and can cause unexpected bugs if not reviewed properly.

Will do. But I admittedly cannot give you a time frame ATM.